buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java

```diff
@@ -0,0 +1,121 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle.reaper;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A standalone process that will reap external services after a build dies.
+ *
+ * <h2>Input</h2>
+ * Since how to reap a given service is platform and service dependent, this tool
+ * operates on system commands to execute. It takes a single argument, a directory
+ * that will contain files with reaping commands. Each line in each file will be
+ * executed with {@link Runtime#getRuntime()#exec}.
+ *
+ * The main method will wait indefinitely on the parent process (Gradle) by
+ * reading from stdin. When Gradle shuts down, whether normally or abruptly, the
+ * pipe will be broken and read will return.
+ *
+ * The reaper will then iterate over the files in the configured directory,
+ * and execute the given commands. If any commands fail, a failure message is
+ * written to stderr. Otherwise, the input file will be deleted. If no inputs
+ * produced errors, the entire input directory is deleted upon completion of reaping.
+ */
+public class Reaper implements Closeable {
+
+    private Path inputDir;
+    private boolean failed;
+
+    private Reaper(Path inputDir) {
+        this.inputDir = inputDir;
+        this.failed = false;
+    }
+
+    public static void main(String[] args) throws Exception {
+        if (args.length != 1) {
+            System.err.println("Expected one argument.\nUsage: java -jar reaper.jar <DIR_OF_REAPING_COMMANDS>");
+            System.exit(1);
+        }
+        Path inputDir = Paths.get(args[0]);
+
+        try (Reaper reaper = new Reaper(inputDir)){
+            System.in.read();
+            reaper.reap();
+        }
+    }
+
+    private void reap() {
+        try {
+            final List<Path> inputFiles = Files.list(inputDir)
+                .filter(p -> p.getFileName().toString().endsWith(".cmd")).collect(Collectors.toList());
+
+            for (Path inputFile : inputFiles) {
+                System.out.println("Process file: " + inputFile);
+                for (String line : Files.readAllLines(inputFile)) {
+                    System.out.println("Running command: " + line);
+                    String[] command = line.split(" ");
```

Comment:
I thought about doing explicit commands, but this makes the reaper much less useful as a general purpose tool, because it must be modified every time there is a new usecase.

> Not dealing with platform specifics hear will mean that he complexities of doing so will leek to all the users of this functionality.

Platforms specifics are a reality. The user of the reaper knows better how to deal with shutting down a service on the current platform than the reaper could.

I chose to make this simple for now. We can always change things like making spaces work, or even allowing multiple lines (eg for trying a SIGTERM followed by SIGKILL if that doesn't work after some timeout), but I don't see a concrete need for such complexity yet.

Reply:
I was on the fence with it too, I'm fine to leave it as it is for now.